### PR TITLE
Refactor CSWRepository to GeoNetworkRepository

### PR DESCRIPTION
**Pull request by KellyStathis**
Background on this change: https://computecanada.atlassian.net/browse/RDM-5223

Because we were not getting the information we needed from Hakai's CSW implementation, I worked on switching us over to using the GeoNetwork API. This returns full ISO 19115 metadata, which is more complete.

Metadata changes resulting from this:

- New fields: subject, affiliation, access, and rights
- Updated fields:
  - authors are now individual (vs. generic "Hakai institute")
  - date is now the actual publication date (unsure what it was before; possibly last updated)
  - item_url has DOI where available
  - keywords include additional keywords that were previously missing

Update repos.json for Hakai institute for testing is on the branch and also here:

```
{
    "repos": [
      {
          "name": "Hakai Institute",
          "type": "geonetwork",
          "url": "https://hecate.hakai.org/geonetwork/srv/api/records/",
          "homepage_url": "https://hecate.hakai.org/",
          "thumbnail": "https://{{ web_hostname }}/discover/img/sources/hakai_80x80.png",
          "item_url_pattern": "https://hecate.hakai.org/geonetwork/srv/eng/catalog.search#/metadata/%id%",
          "repo_registry_uri": "https://www.re3data.org/repository/r3d100013227",
          "enabled": true
      }
    ]
}

```

### CODE REVIEW

**Comment by KellyStathis** [comment]
I should mention Pipfile updates: I can no longer get owslib to work locally, and a bonus of this change is that it is no longer required. I also removed the Python version requirement from the Pipfile - let me know if that causes problems.

File: conf/repos.json [hunk 1]
```diff
@@ -1,4 +1,14 @@
 {
     "repos": [
```
Comment (prairiewest) [inline conf/repos.json]: Probably shouldn't have added this file to the commit

File: Pipfile [hunk 2]
```diff
@@ -24,6 +23,4 @@ psycopg2-binary = "*"
 python-dateutil = "*"
 dateparser = "==1.0.0"
 ftfy = "*"
-
-[requires]
-python_version = "3.7"
+rdflib = "*"
```
Comment (prairiewest) [inline Pipfile:26]: Also update requirements.txt

**Review by prairiewest — COMMENTED** [review]
Don't we want a DB migration script to delete old Hakai data?

I saw this in the log, not sure what it really means:

https://hecate.hakai.org/geonetwork/srv/metadata//registries/vocabularies//concepts/Downloadable Data does not look like a valid URI, trying to serialize this will break.
https://hecate.hakai.org/geonetwork/srv/api/records/94cdfcba-bbd4-4053-8976-75de69460c14/attachments/Sea wrack correction data field attributes.docx.pdf does not look like a valid URI, trying to serialize this will break.

(etc)

File: conf/repos.json (hunk 1 above)
Comment (prairiewest) [reply conf/repos.json]: Or maybe you were going to remove it before merging, if that's the case then carry on :)

**Comment by KellyStathis** [comment]
> Don't we want a DB migration script to delete old Hakai data?
Yes, we should drop all rows from the records table.

I know we don't cascade on delete, but do we periodically clean up rows in crosstables that no longer have an associated record?


> I saw this in the log, not sure what it really means:
> 
> https://hecate.hakai.org/geonetwork/srv/metadata//registries/vocabularies//concepts/Downloadable Data does not look like a valid URI, trying to serialize this will break. https://hecate.hakai.org/geonetwork/srv/api/records/94cdfcba-bbd4-4053-8976-75de69460c14/attachments/Sea wrack correction data field attributes.docx.pdf does not look like a valid URI, trying to serialize this will break.
> 
> (etc)

This is a side effect of parsing the RDF in line 41: https://github.com/frdr-dfdr/frdr_harvest/pull/15/files#diff-c4e3cd679230c4580a7086d0ea8a4e63c39fc6c644f1147c854171cb8255778dR41

These messages don't matter for our purposes - unfortunately, I couldn't figure out how to suppress them from the rdflib documentation. Here is the parse function: https://rdflib.readthedocs.io/en/stable/apidocs/rdflib.html?highlight=parse()#rdflib.graph.Graph.parse

If you see a way to remove them, I support doing so :)

File: Pipfile (hunk 2 above)
Comment (KellyStathis) [reply Pipfile:26]: Thank you!

File: conf/repos.json (hunk 1 above)
Comment (KellyStathis) [reply conf/repos.json]: I shouldn't have committed it, fixed now!

File: harvester/GeoNetworkRepository.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+
+        try:
+            g = Graph()
+            request_success = False
+            request_count = 0
+            while not request_success and request_count < 5:
+                try:
+                    g.parse(self.url)
+                    request_success = True
+                except urllib.error.HTTPError as e:
```
Comment (KellyStathis) [inline harvester/GeoNetworkRepository.py:43]: @prairiewest is this a decent way to handle this, given reliability issues?

File: harvester/GeoNetworkRepository.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        request_count = 0
+        while not request_success and request_count < 5:
+            try:
+                xml_record_url = "https://hecate.hakai.org/geonetwork/srv/api/records/{}/formatters/xml".format(record["local_identifier"])
+                response = requests.request("GET", xml_record_url)
+                if response.status_code == 200:
+                    request_success = True
+                elif response.status_code == 400:
+                    request_count += 1
+                    self.logger.info("Trying again to fetch record {}: Response {}".format(record["local_identifier"], response.status_code))
```
Comment (KellyStathis) [inline harvester/GeoNetworkRepository.py:284]: @prairiewest similar question as above - basically after a few tries (usually on the 2nd or 3rd), I'll get a 200, but there are occasionally a few glitches first.

File: harvester/GeoNetworkRepository.py (hunk 3 above)
Comment (prairiewest) [reply harvester/GeoNetworkRepository.py:43]: Sure. You may want to sleep for longer than 1 second - maybe 3?

File: harvester/GeoNetworkRepository.py (hunk 4 above)
Comment (prairiewest) [reply harvester/GeoNetworkRepository.py:284]: The "not 200" codes may be due to some caching on the server side that has to retrieve data, and can't get it done within the required time... but on later requests, the data is finally available and the request succeeds.  So sleeping longer may cut down on the number of error requests.

Comment (KellyStathis) [reply harvester/GeoNetworkRepository.py:284]: Thanks, this is helpful! I've increased the sleep time.

**Pull request merged**